wsmux/session.go

```diff
@@ -0,0 +1,334 @@
+package wsmux
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+/*
+TODO: Add ping and pong handlers
+*/
+
+const (
+	defaultQueueSize            = 20
+	defaultStreamQueueSize      = 20
+	defaultKeepAliveInterval    = 30 * time.Second
+	defaultStreamAcceptDeadline = 30 * time.Second
+)
+
+// Session implements net.Listener. Allows creating and acception multiplexed streams over ws
+type Session struct {
+	mu       sync.Mutex
+	streams  map[uint32]*stream
+	streamCh chan *stream
+	writes   chan frame
+	conn     *websocket.Conn
+
+	keepAliveInterval    time.Duration
+	streamAcceptDeadline time.Duration
+
+	logger Logger
+
+	nextID uint32
+
+	closed chan struct{} // nil channel
+
+	closeConn bool
+
+	remoteCloseCallback func()
+}
+
+func newSession(conn *websocket.Conn, server bool, conf Config) *Session {
+	s := &Session{}
+	s.conn = conn
+	s.streams = make(map[uint32]*stream)
+	s.streamCh = make(chan *stream, defaultStreamQueueSize)
+	s.writes = make(chan frame, defaultQueueSize)
+
+	s.closed = make(chan struct{})
+	s.closeConn = true
+
+	s.remoteCloseCallback = conf.RemoteCloseCallback
+
+	if conf.KeepAliveInterval == 0 {
```

Comment:
I would suggest that when defining session you do:
```go
s := &Session{
  keepAliveInterval: defaultKeepAliveInterval,
  logger:  &nilLogger{},
  nextID: 0,
  ... // setting defaults
}
```

Then later we can do:
```go
if conf.KeepAliveInterval != 0 {
  s.keepAliveInterval = conf.KeepAliveInterval
}
```
and similar for the other variables...